src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs

```diff
@@ -0,0 +1,593 @@
+        {
+            get { return this.outerChannel; }
+            set { this.outerChannel = value; }
+        }
+
+        IList<Type> IServiceDispatcher.SupportedChannelTypes => throw new NotImplementedException();
+
+        public ServiceHostBase Host =>  InnerServiceDispatcher.Host;
+
+        //private 
+
+        // This method should only be called at Open time, since it looks up the identity based on the 
+        // thread token
+        //void ComputeEndpointIdentity()
+        //{
+        //    EndpointIdentity result = null;
+        //    if (this.State == CommunicationState.Opened)
+        //    {
+        //        if (this.SecurityProtocolFactory != null)
+        //        {
+        //            result = this.SecurityProtocolFactory.GetIdentityOfSelf();
+        //        }
+        //        else if (this.SessionServerSettings != null && this.SessionServerSettings.SessionProtocolFactory != null)
+        //        {
+        //            result = this.SessionServerSettings.SessionProtocolFactory.GetIdentityOfSelf();
+        //        }
+        //    }
+        //    if (result == null)
+        //    {
+        //        result = base.GetProperty<EndpointIdentity>();
+        //    }
+        //    this.identity = result;
+        //}
+
+
+
+        void ThrowIfProtocolFactoryNotSet()
+        {
+            if (this.securityProtocolFactory == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.SecurityProtocolFactoryShouldBeSetBeforeThisOperation)));
+            }
+        }
+        public async Task<IServiceChannelDispatcher> CreateServiceChannelDispatcherAsync(IChannel outerChannel)
+        {
+            this.outerChannel = outerChannel;
+            IChannel securityChannel = GetInnerChannel(outerChannel);
+            var sessionIdleManager = securityChannel.GetProperty<ServiceChannel.SessionIdleManager>();
+            IChannelBinder binder = null;
+            if (securityChannel is IReplyChannel)
+            {
+                var rcbinder = securityChannel.GetProperty<SecurityReplyChannelBinder>();
+                rcbinder.Init(securityChannel as IReplyChannel, BaseAddress);
+                binder = rcbinder;
+            }
+            else
+            {
+                throw new NotImplementedException();
+            }
+            //Open the session server 
+            await this.SessionServerSettings.OnOpenAsync(ServiceDefaults.OpenTimeout);
+
+            // TODO: MUST add the proper messsage
+            if (this.SecurityAuthChannelDispatcher == null)
+                throw new Exception("SecurityAuthChannelDispatcher can't be null");
+
+            ServiceDispatcher securitySessionAuthDispatcher = new ServiceDispatcher(this.SecurityAuthChannelDispatcher);
+            var channelHandler = new ChannelHandler(Binding.MessageVersion, binder, securityChannel.GetProperty<ServiceThrottle>(),
```

Comment:
Need to create a single instance in an initialization method. It then needs optionally needs to be given to the ChannelDemuxer or used as the forward IServiceDispatcher

Reply:
done.. let's review